spinalcordtoolbox/image.py

```diff
@@ -481,54 +501,45 @@ def save(self, path=None, dtype=None, verbose=1, mutable=False):
-        elif path is not None and os.path.isdir(path) and self.absolutepath is not None:
-            # Save to destination directory with original basename
-            path = os.path.join(os.path.abspath(path), os.path.basename(self.absolutepath))
-
-        path = path or self.absolutepath
-
-        if dtype is not None:
-            dst = self.copy()
-            dst.change_type(dtype)
-            data = dst.data
-        else:
-            data = self.data
-
-        # update header
-        hdr = self.hdr.copy() if self.hdr else None
-        if hdr:
-            hdr.set_data_shape(data.shape)
-            # Update dtype if provided (but not if based on SCT-specific values: 'minimize')
-            if (dtype is not None) and (dtype not in ['minimize', 'minimize_int']):
-                hdr.set_data_dtype(dtype)
-
-        # nb. that copy() is important because if it were a memory map, save()
-        # would corrupt it
-        img = nib.nifti1.Nifti1Image(data.copy(), None, hdr)
-        if os.path.isfile(path):
-            if verbose:
-                logger.warning('File ' + path + ' already exists. Will overwrite it.')
-
-        # save file
-        if os.path.isabs(path):
-            logger.debug("Saving image to %s orientation %s shape %s",
-                         path, self.orientation, data.shape)
-        else:
-            logger.debug("Saving image to %s (%s) orientation %s shape %s",
-                         path, os.path.abspath(path), self.orientation, data.shape)
-
-        nib.save(img, path)
-
         if mutable:
+            # do all modifications in-place
+            if path is None:
+                # Use self.absolutepath by default
+                if self.absolutepath is None:
+                    raise ValueError("Don't know where to save the image (no absolutepath or path parameter)")
+                path = self.absolutepath
+            elif os.path.isdir(path):
+                # Save to destination directory with original basename
+                if self.absolutepath is None:
+                    raise ValueError("Don't know where to save the image (no absolutepath and path parameter is dir)")
+                path = os.path.join(os.path.abspath(path), os.path.basename(self.absolutepath))
             self.absolutepath = path
-            self.data = data
-
-        if not os.path.isfile(path):
-            raise RuntimeError("Couldn't save {}".format(path))
-
+            if os.path.isfile(path) and verbose:
+                logger.warning("File %s already exists. Will overwrite it.", path)
+            if os.path.isabs(path):
+                logger.debug("Saving image to %s orientation %s shape %s",
+                             path, self.orientation, self.data.shape)
+            else:
+                logger.debug("Saving image to %s (%s) orientation %s shape %s",
+                             path, os.path.abspath(path), self.orientation, self.data.shape)
```

Comment:
I just realized... we're setting `self.absolutepath` to be `path` without actually... calling `abspath`?

I'm wondering if we'd be better off printing these debug messages, and _then_ setting `self.absolutepath`:

```diff
-           self.absolutepath = path
            if os.path.isfile(path) and verbose:
                logger.warning("File %s already exists. Will overwrite it.", path)
            if os.path.isabs(path):
                logger.debug("Saving image to %s orientation %s shape %s",

                             path, self.orientation, self.data.shape)
            else:
                logger.debug("Saving image to %s (%s) orientation %s shape %s",
                             path, os.path.abspath(path), self.orientation, self.data.shape)
+           self.absolutepath = os.path.abspath(path)	
```

Reply:
Good catch! :+1: